File: examples/javascript/sharded_index_example.mjs
/** ShardedIndex -- HNSW shards with k-means routing and rebalancing. */

import { randomUUID } from "node:crypto";
import { MembrainShardedIndex } from "membrain";

const DIMENSION = 128;

function embed(seed) {
  const out = [];
  let state = seed * 1000 + 1;
  for (let i = 0; i < DIMENSION; i++) {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    out.push((state / 0x7fffffff) * 2 - 1);
  }
  return out;
}

const ids = Array.from({ length: 1000 }, () => randomUUID());
const vectors = Array.from({ length: 1000 }, (_, i) => embed(i));

const config = {
  dimension: DIMENSION,
  num_shards: 4,
  nprobe: 2,
  overlap_factor: 1.5,
  hnsw_config: { m: 16, ef_construction: 100, ef_search: 50 },
};

const index = MembrainShardedIndex.build(ids, vectors, config);

try {
  const info = index.info();
  console.log(`Built: ${index.len()} vectors across ${info.num_shards} shards`);
  for (const shard of info.shards) {
    console.log(`  Shard ${shard.shard_index}: ${shard.count} vectors`);
  }

  const results = index.search(embed(0.5), 5);
  for (const r of results) {
    console.log(`  id=${r.id.slice(0, 8)}... score=${r.score.toFixed(4)}`);
  }

  for (let i = 0; i < 50; i++) index.add(randomUUID(), embed(1000 + i));

  index.rebalance();
  console.log(`\nAfter rebalance: ${index.len()} vectors, stddev=${index.info().size_stddev.toFixed(2)}`);
} finally {
  index.close();
}

File: examples/python/sharded_index_example.py
"""ShardedIndex -- HNSW shards with k-means routing and rebalancing."""

import random
import uuid

from membrain import MembrainShardedIndex

DIMENSION = 128


def embed(seed: float) -> list[float]:
    random.seed(seed)
    return [random.gauss(0, 1) for _ in range(DIMENSION)]


def main() -> None:
    ids = [str(uuid.uuid4()) for _ in range(1000)]
    vectors = [embed(seed=float(i)) for i in range(1000)]

    config = {
        "num_shards": 4,
        "nprobe": 2,
        "overlap_factor": 1.5,
        "hnsw_config": {"m": 16, "ef_construction": 100, "ef_search": 50},
    }

    with MembrainShardedIndex.build(
        dimension=DIMENSION, ids=ids, vectors=vectors, config=config,
    ) as index:
        info = index.info()
        print(f"Built: {len(index)} vectors across {info.num_shards} shards")
        for shard in info.shards:
            print(f"  Shard {shard.shard_index}: {shard.count} vectors")

        results = index.search(embed(seed=0.5), k=5)
        for result in results:
            print(f"  id={result.id[:8]}... score={result.score:.4f}")

        for i in range(50):
            index.add(str(uuid.uuid4()), embed(seed=float(1000 + i)))

        index.rebalance()
        print(f"\nAfter rebalance: {len(index)} vectors, stddev={index.info().size_stddev:.2f}")


if __name__ == "__main__":
    main()
